apm-agent-core/src/main/java/co/elastic/apm/agent/impl/GlobalTracer.java

```diff
@@ -54,17 +63,52 @@ public static ElasticApmTracer requireTracerImpl() {
         return Objects.requireNonNull(getTracerImpl(), "Registered tracer is not an instance of ElasticApmTracer");
     }
 
+    private static void checkClassloader() {
+        ClassLoader cl = GlobalTracer.class.getClassLoader();
+
+        if (classloaderCheck(cl)) {
+            return;
+        }
+
+        throw new IllegalStateException("Agent setup error: must be loaded in bootstap classloader, should not be included in application dependencies. Classloader : " + cl);
+    }
+
+    private static boolean classloaderCheck(@Nullable ClassLoader cl) {
+        if (cl == null) {
+            // agent currently loaded in the bootstrap CL, which is the current correct location
+            return true;
+        }
+
+        // when not in the bootstrap CL, it might be either:
+        // - when executing tests, without a packaged agent
+        // - in production when there is a setup error, with a packaged agent that has a manifest
+        try {
+            Enumeration<URL> resources = cl.getResources("META-INF/MANIFEST.MF");
+            while (resources.hasMoreElements()) {
+                Manifest manifest = new Manifest(resources.nextElement().openStream());
+                Attributes attributes = manifest.getMainAttributes();
+                String preMain = attributes.getValue("Premain-Class");
+                if (preMain != null && preMain.startsWith("co.elastic.apm.agent")) {
+                    return false;
+                }
+            }
+        } catch (IOException e) {
+            throw new IllegalStateException(e);
+        }
```

Comment:
[minor]

Maybe you can reuse https://github.com/elastic/apm-agent-java/blob/f9267ff7fc7df619cb2272b8790fde618055c8f6/apm-agent-core/src/main/java/co/elastic/apm/agent/util/VersionUtils.java#L105

Or even just do something like https://github.com/elastic/apm-agent-java/blob/f9267ff7fc7df619cb2272b8790fde618055c8f6/apm-agent-premain/src/main/java/co/elastic/apm/agent/premain/AgentMain.java#L157 - we only want to see the class is coming from within a jar

Reply:
Very good idea, it's even simpler to check and does not require manifest parsing.